File: java-algorithm/src/com/jun/四数之和_18.java
package com.jun;

import com.jun.util.ArrayUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 给定一个包含 n 个整数的数组 nums 和一个目标值 target，判断 nums 中是否存在四个元素 a，b，c 和 d ，使得 a + b + c + d 的值与 target 相等？找出所有满足条件且不重复的四元组。
 * 注意：
 * 答案中不可以包含重复的四元组。
 * 示例：
 * 给定数组 nums = [1, 0, -1, 0, -2, 2]，和 target = 0。
 * 满足要求的四元组集合为：
 * [
 * [-1,  0, 0, 1],
 * [-2, -1, 1, 2],
 * [-2,  0, 0, 2]
 * ]
 *
 * @author Jun
 * @date 2020/1/11 上午12:44
 */
public class 四数之和_18 {
    public static void main(String[] args) {
//        int[] nums = {-5, -4, -3, -2, -1, 0, 0, 1, 2, 3, 4, 5};
//        int target = 0;

        int[] nums = {1, -2, -5, -4, -3, 3, 3, 5};
        int target = -11;
        ArrayUtil.outputList_2(fourSum(nums, target));
    }

    public static List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> result = new ArrayList<>();
        if (nums == null || nums.length < 4) return new ArrayList<>();
        Arrays.sort(nums);
        int length = nums.length;
        for (int i = 0; i < length - 3; i++) {
            for (int j = i + 1; j < length - 2; ++j) {
                int l = j + 1;
                int r = length - 1;
                while (l < r) {
                    if (nums[i] + nums[j] + nums[l] + nums[r] == target) {
                        result.add(Arrays.asList(nums[i], nums[j], nums[l], nums[r]));
                        while (l < r && nums[l] == nums[l + 1]) l++;
                        l++;
                        while (l < r && nums[r] == nums[r - 1]) r--;
                        r--;
                        continue;
                    }
                    if (nums[i] + nums[j] + nums[l] + nums[r] < target) {
                        while (l < r && nums[l] == nums[l + 1]) l++;
                        l++;
                        continue;
                    }
                    while (l < r && nums[r] == nums[r - 1]) r--;
                    r--;
                }
                while (j < length - 2 && nums[j] == nums[j + 1]) j++;
            }
            while (i < length - 3 && nums[i] == nums[i + 1]) i++;
        }
        return result;
    }
}

File: python-algorithm/四数之和_18/四数之和_18.py
class 四数之和_18:
    def four_sum(self, nums, target):
        def find_n_sum(nums, target, n, result, results):
            if len(nums) < n or n < 2 or target < nums[0] * n or target > nums[-1] * n:
                return
            if n == 2:
                l, r = 0, len(nums) - 1
                while l < r:
                    s = nums[l] + nums[r]
                    if s == target:
                        results.append(result + [nums[l], nums[r]])
                        while l < r and nums[l] == nums[l + 1]:
                            l += 1
                        l += 1
                        while l < r and nums[r] == nums[r - 1]:
                            r -= 1
                        r -= 1
                    elif s < target:
                        l += 1
                    else:
                        r -= 1
            else:
                for i in range(len(nums) - n + 1):
                    if i == 0 or (i > 0 and nums[i - 1] != nums[i]):
                        find_n_sum(nums[i + 1], target - nums[i], n - 1, result + [nums[i], result])

        results = []
        find_n_sum(sorted(nums), target, 4, [], results)
        return results
